numpy/random/mtrand/mtrand.pyx

```diff
@@ -964,20 +964,26 @@ cdef class RandomState:
             high = low
             low = 0
 
+        # '_randint_type' is defined in
+        # 'generate_randint_helpers.py'
         key = np.dtype(dtype).name
-        if not key in _randint_type:
+        if key not in _randint_type:
             raise TypeError('Unsupported dtype "%s" for randint' % key)
+
         lowbnd, highbnd, randfunc = _randint_type[key]
 
-        if low < lowbnd:
+        ilow = int(low)
+        ihigh = int(high)
+
+        if ilow < lowbnd:
             raise ValueError("low is out of bounds for %s" % (key,))
-        if high > highbnd:
+        if ihigh > highbnd:
             raise ValueError("high is out of bounds for %s" % (key,))
-        if low >= high:
+        if ilow >= ihigh:
             raise ValueError("low >= high")
 
         with self.lock:
-            ret = randfunc(low, high - 1, size, self.state_address)
+            ret = randfunc(ilow, ihigh - 1, size, self.state_address)
```

Comment:
Not sure what you meant by `# correct but bad for randint`. You're in for a bad time if you ask for an upper bound that doesn't come close to fitting in the container type. What does passing `dtype=None` do normally?